src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/GlobalFlowStateAnalysis/GlobalFlowStateAnalysisValueSet.cs

```diff
@@ -40,17 +40,49 @@ internal sealed class GlobalFlowStateAnalysisValueSet : CacheBasedEquatable<Glob
             Kind = kind;
         }
 
-        public GlobalFlowStateAnalysisValueSet(IAbstractAnalysisValue analysisValue)
-            : this(ImmutableHashSet.Create(analysisValue), ImmutableHashSet<GlobalFlowStateAnalysisValueSet>.Empty, height: 0, GlobalFlowStateAnalysisValueSetKind.Known)
+        public static GlobalFlowStateAnalysisValueSet Create(
+            ImmutableHashSet<IAbstractAnalysisValue> analysisValues,
+            ImmutableHashSet<GlobalFlowStateAnalysisValueSet> parents,
+            int height)
         {
+            Debug.Assert(!analysisValues.IsEmpty || !parents.IsEmpty);
+            return new(analysisValues, parents, height, GlobalFlowStateAnalysisValueSetKind.Known);
         }
 
-        public GlobalFlowStateAnalysisValueSet(GlobalFlowStateAnalysisValueSet parent1, GlobalFlowStateAnalysisValueSet parent2)
-            : this(ImmutableHashSet<IAbstractAnalysisValue>.Empty,
-                   ImmutableHashSet.Create(parent1, parent2),
-                   height: Math.Max(parent1.Height, parent2.Height) + 1,
-                   GlobalFlowStateAnalysisValueSetKind.Known)
+        public static GlobalFlowStateAnalysisValueSet Create(IAbstractAnalysisValue analysisValue)
+            => new(ImmutableHashSet.Create(analysisValue), ImmutableHashSet<GlobalFlowStateAnalysisValueSet>.Empty, height: 0, GlobalFlowStateAnalysisValueSetKind.Known);
+
+        public static GlobalFlowStateAnalysisValueSet CreateWithParents(GlobalFlowStateAnalysisValueSet parent1, GlobalFlowStateAnalysisValueSet parent2)
         {
+            Debug.Assert(parent1.Kind == GlobalFlowStateAnalysisValueSetKind.Known);
+            Debug.Assert(parent2.Kind == GlobalFlowStateAnalysisValueSetKind.Known);
+
+            if (Equals(parent1, parent2))
+            {
+                return parent1;
+            }
+
+            if (parent1.Height == 0 && parent1.AnalysisValues.IsSubsetOf(parent2.AnalysisValues))
+            {
+                return parent2;
+            }
+
+            if (parent2.Height == 0 && parent2.AnalysisValues.IsSubsetOf(parent1.AnalysisValues))
+            {
+                return parent1;
+            }
+
+            if (parent1.Height == parent2.Height &&
+                parent1.AnalysisValues.Count == parent2.AnalysisValues.Count &&
+                Equals(parent1, parent2.GetNegatedValue()))
+            {
+                return Empty;
```

Comment:
❔ Why this path?

Reply:
This was leading to unnecessary analysis iterations for the added test cases

Reply:
Please add a comment on the return statement with a description of what `parent1` and `parent2` might represent in order to return this, and why `Empty` is the right combined result.

Reply:
Sure, will add.